pkg/rollout/revision.go

```diff
@@ -40,13 +41,30 @@ func DetectStableRevisionName(svc *run.Service) string {
 
 // DetectCandidateRevisionName attempts to deduce what revision could be
 // considered a candidate.
-func DetectCandidateRevisionName(svc *run.Service, stable string) string {
+//
+// It also determines if the candidate is different from a previous rollout
+// process. Knowing if a candidate is a new one is useful since metrics cannot
+// be obtained on it (it has 0 traffic), so the rollout process should add some
+// initial traffic to the new revision.
+func DetectCandidateRevisionName(svc *run.Service, stable string) (string, bool) {
 	latestRevision := svc.Status.LatestReadyRevisionName
 	if stable == latestRevision {
-		return ""
+		return "", false
+	}
+
+	// If the latestRevision has previously been treated as a candidate and
+	// failed to meet health checks, no candidate exists.
+	if latestRevision == svc.Metadata.Annotations[LastFailedCandidateRevisionAnnotation] {
+		return "", false
+	}
+
+	isNewCandidate := false
+	previousCandidate := findRevisionWithTag(svc, CandidateTag)
+	if previousCandidate != latestRevision {
+		isNewCandidate = true
 	}
```

Comment:
to determine if a rev is a "new candidate", look at revName==latestReadyRevision

Reply:
I forgot about this when we were talking, but if the traffic to a revision is 0, the revision is (often) not included in the traffic configuration. So we will need to see if the candidate's revision name is included in the traffic config.

If it is not, then the candidate is a new one.
If it is included, just check that Percent > 0.